File: 2022/day07/day07.go
package main

import (
	"fmt"
	"strconv"
	"strings"

	"github.com/icemanblues/advent-of-code/pkg/util"
)

const (
	dayNum   = "07"
	dayTitle = "No Space Left On Device"
)

type File struct {
	name     string
	size     int
	isFile   bool
	parent   *File
	children map[string]*File
}

func NewFile(name string, size int, parent *File) *File {
	return &File{name, size, true, parent, nil}
}

func NewDir(name string, parent *File) *File {
	return &File{name, 0, false, parent, make(map[string]*File)}
}

func (f *File) AddFile(add *File) {
	f.children[add.name] = add
}

type FileSystem struct {
	root *File
	curr *File
}

func NewFS() FileSystem {
	root := NewDir("/", nil)
	return FileSystem{root, root}
}

func parse(filename string) FileSystem {
	input, _ := util.ReadInput(filename)
	fs := NewFS()
	var cmd string
	for _, line := range input {
		fields := strings.Fields(line)
		if fields[0] == "$" {
			cmd = fields[1]

			if cmd == "cd" {
				// handle built-ins for cd
				if fields[2] == ".." {
					fs.curr = fs.curr.parent
				} else if fields[2] == "/" {
					fs.curr = fs.root
				} else {
					fs.curr = fs.curr.children[fields[2]]
				}
			}
			continue
		}

		// must be the output of ls
		if fields[0] == "dir" {
			dir := NewDir(fields[1], fs.curr)
			fs.curr.AddFile(dir)
			continue
		}
		s, _ := strconv.Atoi(fields[0])
		f := NewFile(fields[1], s, fs.curr)
		fs.curr.AddFile(f)
	}
	return fs
}

func sizeof(file *File, memo map[*File]int) map[*File]int {
	if _, ok := memo[file]; ok {
		return memo
	}

	if file.isFile {
		memo[file] = file.size
		return memo
	}

	s := 0
	for _, f := range file.children {
		memo = sizeof(f, memo)
		s += memo[f]
	}
	memo[file] = s
	return memo
}

func main() {
	fmt.Printf("Day %v: %v\n", dayNum, dayTitle)
	fs := parse("input.txt")
	memo := make(map[*File]int)
	memo = sizeof(fs.root, memo)

	dirSize := 0
	for f, s := range memo {
		if !f.isFile && s <= 100000 {
			dirSize += s
		}
	}
	fmt.Printf("Part 1: %v\n", dirSize)

	unused := 70000000 - memo[fs.root]
	need := 30000000
	target := need - unused
	minSize := memo[fs.root]
	for f, s := range memo {
		if !f.isFile && s >= target && s < minSize {
			minSize = s
		}
	}
	fmt.Printf("Part 2: %v\n", minSize)
}

File: 2022/day07/Day07.java
import java.io.*;
import java.util.*;

class Day07 {
    public static class File {
        private String name;
        private int size;
        private boolean isFile;
        private File parent;
        private Map<String, File> children;

        public File(String name, int size, boolean isFile, File parent) {
            this.name = name;
            this.size = size;
            this.isFile = isFile;
            this.parent = parent;
            this.children = new HashMap<>();
        }

        public void addFile(File f) {
            children.put(f.name, f);
        }
    }

    public static void sizeOf(File file, Map<File, Integer> memo) {
        if(memo.containsKey(file)) {
            return;
        }

        if(file.isFile) {
            memo.put(file, file.size);
            return;
        }

        int sum = 0;
        for(File f: file.children.values()) {
            sizeOf(f, memo);
            sum += memo.get(f);
        }
        memo.put(file, sum);
    }

    public static void main(String[] args) throws Exception {
        System.out.println("2022 day 07");

        File root = new File("/", 0, false, null);
        File curr = root;

        try(BufferedReader br = new BufferedReader(new FileReader("test.txt"))) {
            for(String line = br.readLine(); line != null; line = br.readLine()) {
                String[] fields = line.split("\s");
                if("$".equals(fields[0])) {
                    String cmd = fields[1];
                    if("cd".equals(cmd)) {
                        switch(fields[2]) {
                            case "..":
                                curr = curr.parent;
                                break;
                            case "/":
                                curr = root;
                                break;
                            default:
                                curr = curr.children.get(fields[2]);
                        }

                    }
                    continue;
                }

                // must be list output
                if("dir".equals(fields[0])) {
                    File dir = new File(fields[1], 0, false, curr);
                    curr.addFile(dir);
                } else {
                    File file = new File(fields[1], Integer.parseInt(fields[0]), true, curr);
                    curr.addFile(file);
                }
            }
        }

        // use memoization to aggregate the size
        Map<File, Integer> memo = new HashMap<>();
        sizeOf(root, memo);
        
        // part 1
        int sum = 0;
        for(File f: memo.keySet()) {
            int s = memo.get(f);
            if(!f.isFile && s <= 100000) {
                sum += memo.get(f);
            }
        }
        System.out.println("Part 1: "+sum);

        int unused = 70000000 - memo.get(root);
        int target = 30000000 - unused;
        int min = memo.get(root);
        for(File f: memo.keySet()) {
            int s = memo.get(f);
            if(!f.isFile && s >= target && s < min) {
                min = s;
            }
        }
        System.out.println("Part 2: "+min);
    }
}
